Python 3.13 support

Deprecated function PyEval_ThreadsInitialized() was removed in Python 3.13. 
pip install cpuid will fail the compilation of cpuid-native. 
PyTorch 2.6 is now supporting 3.13. cpuid for 3.13 will be useful.


$ pip install --upgrade cpuid
Collecting cpuid
  Using cached cpuid-0.0.11-py3-none-any.whl
Collecting cpuid-native>=0.0.8 (from cpuid)
  Using cached cpuid_native-0.0.8.tar.gz (11 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: cpuid-native
  Building wheel for cpuid-native (pyproject.toml) ... error
  error: subprocess-exited-with-error

  × Building wheel for cpuid-native (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [19 lines of output]
      running bdist_wheel
      running build
      running build_ext
      building 'cpuid_native' extension
      creating build/temp.linux-x86_64-cpython-313
      g++ -pthread -B /datadisk/xfeng8/miniforge3/envs/python313/shared/python_compiler_compat -fno-strict-overflow -Wsign-compare -DNDEBUG -O2 -Wall -fPIC -O2 -isystem /datadisk/xfeng8/miniforge3/envs/python313/include -fPIC -O2 -isystem /datadisk/xfeng8/miniforge3/envs/python313/include -fPIC -I/datadisk/xfeng8/miniforge3/envs/python313/include/python3.13 -c cpuidmodule.cpp -o build/temp.linux-x86_64-cpython-313/cpuidmodule.o
      cpuidmodule.cpp: In function ‘PyObject* PyInit_cpuid_native()’:
      cpuidmodule.cpp:129:11: error: ‘PyEval_ThreadsInitialized’ was not declared in this scope
        129 |     if (! PyEval_ThreadsInitialized()) {
            |           ^~~~~~~~~~~~~~~~~~~~~~~~~
      cpuidmodule.cpp:130:27: warning: ‘void PyEval_InitThreads()’ is deprecated [-Wdeprecated-declarations]
        130 |         PyEval_InitThreads();
            |         ~~~~~~~~~~~~~~~~~~^~
      In file included from /datadisk/xfeng8/miniforge3/envs/python313/include/python3.13/Python.h:124,
                       from cpuidmodule.cpp:5:
      /datadisk/xfeng8/miniforge3/envs/python313/include/python3.13/ceval.h:114:37: note: declared here
        114 | Py_DEPRECATED(3.9) PyAPI_FUNC(void) PyEval_InitThreads(void);
            |                                     ^~~~~~~~~~~~~~~~~~
      error: command '/usr/bin/g++' failed with exit code 1
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for cpuid-native
Failed to build cpuid-native
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (cpuid-native)

--- a/cpuidmodule.cpp
+++ b/cpuidmodule.cpp
@@ -124,11 +124,15 @@ initcpuid_native(void)
 #endif /* PY_MAJOR_VERSION >= 3 */
 
 {
+    // Not necessary after Python 3.7
+    // See: https://docs.python.org/3.12/c-api/init.html#c.PyEval_ThreadsInitialized
+#if (PY_MAJOR_VERSION < 3) || ((PY_MAJOR_VERSION == 3) && (PY_MINOR_VERSION < 7))
     // Make sure the GIL has been created since we need to acquire it in our
     // callback to safely call into the python application.
     if (! PyEval_ThreadsInitialized()) {
         PyEval_InitThreads();
     }
+#endif
 
 
 #if PY_MAJOR_VERSION >= 3

